### PR TITLE
[SofaMatrix] Eigen can be used to reorder a mesh in order to reduce fill-in

### PR DESCRIPTION
**Pull request by alxbilger**
The component FillReducingOrdering, introduced in https://github.com/sofa-framework/sofa/pull/2501, was using the METIS library to reorder the vertices. Now, it is able to use approximate minimum degree algorithm from the Eigen library.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Comment by alxbilger** [comment]
Because I know the SOFA community likes screenshots a lot:

image


Left image: the original matrix, without any permutation
Right image: permuted matrix

When the matrix is permuted, the lower matrix of the LU factorization has 80% less non-zero values than when the matrix is not permuted. Hence the speed up!

File: applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl [hunk 1]
```diff
@@ -176,4 +221,24 @@ void FillReducingOrdering<DataTypes>::updateElements(
...
+template <class DataTypes>
+void FillReducingOrdering<DataTypes>::updateMesh()
+{
+    // Update topology
+    updateElements(l_topology->getHexahedra(), d_hexahedra);
+    updateElements(l_topology->getTetrahedra(), d_tetrahedra);
+
+    // Update mechanical state
+    auto pos = helper::getWriteOnlyAccessor(d_position);
+    pos.clear();
```
Comment (damienmarchal) [inline applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl]: I wonder if pos.resize(previousPos.size() then pos[0] is very different or equal in overhead?

File: applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl [hunk 2]
```diff
@@ -23,13 +23,16 @@
...
+#include <Eigen/Sparse>
+
 namespace sofa::component::linearsolver
 {
 
 template <class DataTypes>
 FillReducingOrdering<DataTypes>::FillReducingOrdering()
     : l_mstate(initLink("mstate", "Mechanical state to reorder"))
     , l_topology(initLink("topology", "Topology to reorder"))
+    , d_orderingMethod(initData(&d_orderingMethod, sofa::helper::OptionsGroup(2, "Metis", "AMD"), "orderingMethod", "Ordering method. AMD means Approximate minimum degree."))
```
Comment (damienmarchal) [inline applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl]: If AMD means for something... is there a possible hint for the METIS one ?

File: applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl [hunk 3]
```diff
@@ -41,6 +44,8 @@ FillReducingOrdering<DataTypes>::FillReducingOrdering()
     addOutput(&d_position);
     addOutput(&d_hexahedra);
     addOutput(&d_tetrahedra);
+
+    sofa::helper::getWriteAccessor(d_orderingMethod)->setSelectedItem(0);
```
Comment (damienmarchal) [inline applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl:51]: I wonder if there is no room for a PR to make this writing a bit more elegant (not in this PR).

File: applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl (hunk 2 above)
Comment (alxbilger) [reply applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl]: Metis is the library

File: applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl (hunk 1 above)
Comment (alxbilger) [reply applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl]: you are right. Usually, I `reserve` before calling the `push_back`'s, which is equivalent to your suggestion.

File: applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl (hunk 2 above)
Comment (alxbilger) [reply applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl]: I've changed the options. Let me know what you think

File: applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl [hunk 4]
```diff
@@ -23,13 +23,19 @@
 
 #include <SofaMatrix/FillReducingOrdering.h>
 
+#include <Eigen/Sparse>
```
Comment (guparan) [inline applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl:26]: Eigen dependency should be visible in CMakeLists.txt and Config.cmake.in

**Review by guparan — CHANGES_REQUESTED** [review]

File: applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl (hunk 4 above)
Comment (alxbilger) [reply applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl:26]: Done, even if Eigen was already a dependency of another module.

How can I test a Config.cmake.in script?

Comment (guparan) [reply applications/plugins/SofaMatrix/src/SofaMatrix/FillReducingOrdering.inl:26]: We don't have such test in our CI. 
Currently the best way is to generate binaries with the command `[ci-build][generate-binaries]` and then try to build a plugin on top of those binaries.

**Comment by guparan** [comment]
[ci-build][with-all-tests][generate-binaries]

**Review by guparan — APPROVED** [review]

**Pull request merged**